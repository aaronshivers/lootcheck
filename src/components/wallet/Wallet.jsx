import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { makeDeposit, makeWithdrawal } from '../../actions/balance';

export class Wallet extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      balance: undefined,
    };
  }

  updateBalance = (event) => {
    this.setState({ balance: Number(event.target.value) });
  };

  handleDeposit = () => {
    const { makeDeposit: makeDepositFromProps } = this.props;
    const { balance } = this.state;
    return makeDepositFromProps(balance);
  }

  handleWithdrawal = () => {
    const { makeWithdrawal: makeWithdrawalFromProps } = this.props;
    const { balance } = this.state;
    return makeWithdrawalFromProps(balance);
  }

  render() {
    const { balance } = this.props;

    return (
      <div>
        <h2 className="balance">
          {['Wallet Balance: ', balance]}
        </h2>
        <br />
        <input type="text" className="input-wallet" onChange={this.updateBalance} />
        <button type="button" className="btn-deposit" onClick={this.handleDeposit}>
          Deposit
        </button>
        <button type="button" className="btn-withdrawal" onClick={this.handleWithdrawal}>
          Withdrawal
        </button>
      </div>
    );
  }
}

Wallet.propTypes = {
  balance: PropTypes.number.isRequired,
  makeDeposit: PropTypes.func.isRequired,
  makeWithdrawal: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({ balance: state.balance });
const mapDispatchToProps = { makeDeposit, makeWithdrawal };

export default connect(mapStateToProps, mapDispatchToProps)(Wallet);
