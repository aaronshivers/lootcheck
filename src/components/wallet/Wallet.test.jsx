import { shallow } from 'enzyme';
import React from 'react';
import { Wallet } from './Wallet';

describe('Wallet', () => {
  const mockDeposit = jest.fn();
  const mockWithdrawal = jest.fn();
  const balance = 20;

  const wallet = shallow(
    <Wallet
      balance={balance}
      makeDeposit={mockDeposit}
      makeWithdrawal={mockWithdrawal}
    />,
  );

  it('should render correctly', () => {
    expect(wallet).toMatchSnapshot();
  });

  it('should display the balance from props', () => {
    expect(wallet.find('.balance').text())
      .toEqual('Wallet Balance: 20');
  });

  it('should create an input to deposit into, or withdrawal from the balance',
    () => expect(wallet.exists('.input-wallet')).toBeTruthy());

  describe('when the user types into the wallet input', () => {
    const userBalance = '25';

    beforeEach(() => {
      wallet
        .find('.input-wallet')
        .simulate('change',
          {
            target: {
              value: userBalance,
            },
          });
    });

    it(
      'should update the local wallet balance in `state` and converts it into a number',
      () => {
        expect(wallet.state().balance).toEqual(Number(userBalance));
      },
    );

    describe('and the user wants to make a deposit', () => {
      beforeEach(() => {
        wallet.find('.btn-deposit').simulate('click');
      });

      it(
        'should dispatch the `deposit()` it receives from props with the local balance',
        () => {
          expect(mockDeposit).toHaveBeenCalledWith(Number(userBalance));
        },
      );
    });

    describe('and the user wants to make a withdrawal', () => {
      beforeEach(() => {
        wallet.find('.btn-withdrawal').simulate('click');
      });

      it(
        'should dispatch the `withdrawal()` it receives from props with the local balance',
        () => {
          expect(mockWithdrawal).toHaveBeenCalledWith(Number(userBalance));
        },
      );
    });
  });
});
