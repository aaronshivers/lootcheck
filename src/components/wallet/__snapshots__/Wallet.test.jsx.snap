// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet should render correctly 1`] = `
<div>
  <h2
    className="balance"
  >
    Wallet Balance: 
    20
  </h2>
  <br />
  <input
    className="input-wallet"
    onChange={[Function]}
    type="text"
  />
  <button
    className="btn-deposit"
    onClick={[Function]}
    type="button"
  >
    Deposit
  </button>
  <button
    className="btn-withdrawal"
    onClick={[Function]}
    type="button"
  >
    Withdrawal
  </button>
</div>
`;
